File: ios/MunderMobile/Sources/OfficeStore.swift
import Foundation
import CryptoKit
import Combine

/// Everything the screens show and do, in one place.
@MainActor
public final class OfficeStore: ObservableObject {
    public enum Phase: Equatable {
        case unpaired
        case waiting(code: String)
        case paired
    }

    @Published public private(set) var phase: Phase = .unpaired
    @Published public private(set) var office: PairedOffice?
    @Published public private(set) var overview: Overview?
    @Published public private(set) var peers: [Peer]?
    @Published public private(set) var online = true
    @Published public private(set) var offlineReason: String?
    @Published public var pairError: String?
    @Published public var toast: String?
    @Published public var busy = false
    /// nil = the office this phone is paired with; otherwise a paired office's id, seen through it.
    @Published public private(set) var viewing: String?
    /// The Panel stratum, when this phone was granted the machine. nil + `panelDenied`
    /// means the host has not been granted, which is a locked door, not a failure.
    @Published public private(set) var panel: PanelState?
    @Published public private(set) var panelDenied: String?
    /// The agent the composer writes to ("god" = Michael). Only for this office.
    @Published public var recipient = "god"

    /// The owner check before anything that changes the office (see AppLock).
    /// Defaults to "allowed" so the tests and demo mode need no Face ID.
    public var authorize: (SensitiveAction) async -> Bool = { _ in true }

    private var client: RemoteClient?
    private let defaults: UserDefaults

    private enum K {
        static let office = "munder.office"
        static let pendingCode = "munder.pendingCode"
        static let sessionKey = "session-key"
    }

    public init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        guard let data = defaults.data(forKey: K.office),
              let office = try? JSONDecoder().decode(PairedOffice.self, from: data),
              let raw = Keychain.load(account: K.sessionKey) else { return }
        adopt(office, key: SymmetricKey(data: raw))
        if let code = defaults.string(forKey: K.pendingCode) { phase = .waiting(code: code) } else { phase = .paired }
    }

    private func adopt(_ office: PairedOffice, key: SymmetricKey) {
        let c = RemoteClient(office: office, key: key)
        c.onOfficeChanged = { [weak self] next in
            Task { @MainActor in self?.persist(next) }
        }
        client = c
        self.office = office
    }

    private func persist(_ office: PairedOffice) {
        self.office = office
        if let data = try? JSONEncoder().encode(office) { defaults.set(data, forKey: K.office) }
    }

    // MARK: pairing

    public func pair(address input: String, deviceName: String) async {
        pairError = nil
        guard let address = Address.normalize(input) else {
            pairError = "Escribe la dirección de tu computadora, por ejemplo 192.168.1.64 o 100.101.4.7."
            return
        }
        busy = true
        defer { busy = false }
        do {
            let name = deviceName.trimmingCharacters(in: .whitespaces).isEmpty ? "iPhone" : deviceName
            let r = try await RemoteClient.pair(address: address, deviceName: name)
            try Keychain.save(r.key.withUnsafeBytes { Data($0) }, account: K.sessionKey)
            persist(r.office)
            defaults.set(r.code, forKey: K.pendingCode)
            adopt(r.office, key: r.key)
            phase = .waiting(code: r.code)
        } catch {
            pairError = error.localizedDescription
        }
    }

    /// Poll `hello` until a human accepts the code on the computer.
    public func waitForAccept() async {
        while case .waiting = phase, !Task.isCancelled {
            do {
                let hello: Hello = try await requireClient().call("hello")
                client?.learn(hello.addresses ?? [])
                defaults.removeObject(forKey: K.pendingCode)
                phase = .paired
                toast = "¡Listo! Celular emparejado"
                await refresh()
                return
            } catch let e as RemoteError where e.code == "unknown_device" {
                pairError = "La computadora ya no tiene esta solicitud (caduca a los 10 minutos, o se rechazó). Empieza de nuevo."
                return
            } catch {
                // waiting / not reachable yet: keep polling
            }
            try? await Task.sleep(nanoseconds: 2_000_000_000)
        }
    }

    /// `-MunderDemo`: show the bundled demo office (for simulator screenshots). No network, no Keychain.
    public func loadDemo(overview: Data, peers: Data, panel: Data? = nil) {
        client = nil
        office = PairedOffice(officeId: "fa801f7ab6693f03", name: "michael-victus", boxPub: "", deviceId: "9ed00cd88840bed1",
                              deviceName: "iPhone de Danny", addresses: ["192.168.1.64:47831", "100.101.4.7:47831"],
                              via: ["192.168.1.64:47831": "lan", "100.101.4.7:47831": "tailscale"])
        self.overview = try? RemoteClient.decoder.decode(Overview.self, from: overview)
        self.peers = (try? RemoteClient.decoder.decode(PeersReply.self, from: peers))?.peers
        // The demo host, from the same fixture the tests decode. No authority dance
        // in a screenshot: it is a picture of the tab, not of a real grant.
        if let panel { self.panel = try? RemoteClient.decoder.decode(PanelState.self, from: panel) }
        online = true
        phase = .paired
    }

    public func forget() {
        viewing = nil
        recipient = "god"
        Keychain.delete(account: K.sessionKey)
        defaults.removeObject(forKey: K.office)
        defaults.removeObject(forKey: K.pendingCode)
        client = nil
        office = nil
        overview = nil
        peers = nil
        pairError = nil
        panel = nil
        panelDenied = nil
        phase = .unpaired
    }

    // MARK: the office

    /// Switch the whole app to another office (nil = this one).
    public func view(office id: String?) async {
        guard id != viewing else { return }
        viewing = id
        recipient = "god"
        overview = nil
        await refresh()
    }

    public func refresh() async {
        guard phase == .paired, let c = client else { return }
        let want = viewing
        do {
            let next = try await c.call("overview", args: want.map { ["office": $0] } ?? [:], as: Overview.self)
            guard want == viewing else { return } // switched while this was in flight
            overview = next
            online = true
            offlineReason = nil
        } catch let e as RemoteError {
            online = false
            switch e.code {
            case "unknown_device": offlineReason = "La oficina ya no reconoce este celular (lo olvidaron en la computadora). Olvídala aquí y vuelve a emparejar."
            case "stale": offlineReason = "La hora del celular y la de la computadora no coinciden. Activa la hora automática en los dos."
            default: offlineReason = e.localizedDescription
            }
        } catch {
            online = false
            offlineReason = error.localizedDescription
        }
    }

    /// Ask the office where else it can be reached (its Tailscale IP, a new LAN IP).
    public func refreshAddresses() async {
        guard phase == .paired, let c = client else { return }
        if let hello: Hello = try? await c.call("hello") { c.learn(hello.addresses ?? []) }
    }

    public func refreshPeers() async {
        guard let c = client else { return } // demo mode keeps its bundled peers
        if let r: PeersReply = try? await c.call("peers") { peers = r.peers } else if peers == nil { peers = [] }
    }

    // MARK: the Panel stratum

    /// Ask the host how it is. `no_authority` is the expected answer until a human
    /// runs `munder link panel <celular>` on the machine, so it is kept apart from
    /// the offline banner: the phone is fine, it just was not given the keys.
    public func refreshPanel() async {
        guard phase == .paired, let c = client else { return }
        do {
            panel = try await c.call("panel.state", as: PanelState.self)
            panelDenied = nil
        } catch let e as RemoteError where e.code == "no_authority" {
            panel = nil
            panelDenied = e.localizedDescription
        } catch {
            // The office is reachable but the panel is not: keep the last state.
        }
    }

    /// Press one of the host's buttons. Every one of them can stop your Munder, so
    /// the owner check is not optional here: it is the same lock the other
    /// write-actions use, plus the 30 s window that lets you fix a typo.
    @discardableResult
    public func panelAction(_ action: String, args: [String: String] = [:]) async -> Bool {
        guard await authorize(.panel) else { return false }
        return await send { c in
            let r: PanelActionReply = try await c.call("panel.action", args: ["action": action, "args": args])
            return r.text.isEmpty ? (r.ok ? "Listo" : "No se pudo") : r.text
        }
    }

    /// Two buttons are desktop-only by design (see PANEL_OFF in lib-remote.cjs):
    /// the phone must not install a desktop shortcut, nor widen its own authority.
    public static let panelRemoteBlocked: Set<String> = ["shortcut.install", "link.phoneAuthority"]

    public func answer(_ task: TaskItem, text: String) async -> Bool {
        guard let q = task.question?.q else { return false }
        guard await authorize(.answer) else { return false }
        return await send { c in
            let _: AnswerReply = try await c.call("answer", args: ["task_id": task.id, "q": q, "text": text])
            return "Respuesta enviada a Michael"
        }
    }

    /// This office: to Michael or one agent. Another office: a task for ITS Michael.
    public func ask(_ text: String) async -> Bool {
        if let peer = viewing { return await delegate(to: peer, text: text) }
        guard await authorize(.ask) else { return false }
        let to = recipient
        return await send { c in
            var args: [String: Any] = ["text": text]
            if to != "god" { args["agent"] = to }
            let r: MessageReply = try await c.call("ask", args: args)
            return "Enviado a \(r.name ?? "Michael")"
        }
    }

    public func delegate(to peer: String, text: String) async -> Bool {
        guard await authorize(.delegate) else { return false }
        return await send { c in
            let r: DelegateReply = try await c.call("delegate", args: ["office": peer, "text": text])
            return "Delegada a \(r.office)"
        }
    }

    /// Add an address by hand (say, the Tailscale IP) and prove it reaches THIS office.
    public func addAddress(_ input: String) async {
        guard let c = client, let office = office, let address = Address.normalize(input) else {
            toast = "Esa dirección no se entiende"
            return
        }
        guard await authorize(.addAddress) else { return }
        let probe = RemoteClient(office: PairedOffice(officeId: office.officeId, name: office.name, boxPub: office.boxPub,
                                                      deviceId: office.deviceId, deviceName: office.deviceName, addresses: [address]),
                                 key: sessionKey() ?? SymmetricKey(size: .bits256))
        do {
            let _: Hello = try await probe.call("hello")
            c.add(address: address, via: Address.looksTailscale(address) ? "tailscale" : "manual")
            toast = "Dirección agregada"
        } catch {
            toast = "Esa dirección no contestó como tu oficina: \(error.localizedDescription)"
        }
    }

    public func removeAddress(_ address: String) async {
        guard client != nil, (office?.addresses.count ?? 0) > 1 else { return }
        guard await authorize(.removeAddress) else { return }
        client?.remove(address: address)
    }

    /// Forget a paired office, after the owner confirms it is them. (Abandoning a
    /// pairing that was never accepted — `forget()` from the code screen — needs
    /// no check: nothing is reachable yet.)
    public func forgetAuthorized() async {
        guard await authorize(.forget) else { return }
        forget()
    }

    private func sessionKey() -> SymmetricKey? {
        Keychain.load(account: K.sessionKey).map { SymmetricKey(data: $0) }
    }

    private func requireClient() throws -> RemoteClient {
        guard let c = client else { throw RemoteError.notPaired }
        return c
    }

    /// Run one action; the toast says what happened, success or the office's own words.
    private func send(_ work: (RemoteClient) async throws -> String) async -> Bool {
        guard let c = client else { return false }
        busy = true
        defer { busy = false }
        do {
            toast = try await work(c)
            await refresh()
            return true
        } catch {
            toast = error.localizedDescription
            if (error as? RemoteError)?.code == "question_changed" { await refresh() }
            return false
        }
    }
}

File: android/MunderMobile/app/src/main/java/mx/isyco/munder/mobile/OfficeStore.kt
package mx.isyco.munder.mobile

import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.serialization.encodeToString

/**
 * Todo lo que muestran y hacen las pantallas, en un solo lugar.
 * Espejo de Sources/OfficeStore.swift: misma Phase, mismos ops, mismos
 * toasts en español, mismo panelRemoteBlocked.
 *
 * `secure` nulo = sin persistencia (tests y capturas): la llave y la oficina
 * viven solo en memoria. La app real siempre lo pasa.
 */
class OfficeStore(
    private val secure: SecureStore?,
) : ViewModel() {

    sealed interface Phase {
        data object Unpaired : Phase
        data class Waiting(val code: String) : Phase
        data object Paired : Phase
    }

    private val _phase = MutableStateFlow<Phase>(Phase.Unpaired)
    val phase: StateFlow<Phase> = _phase.asStateFlow()

    private val _office = MutableStateFlow<PairedOffice?>(null)
    val office: StateFlow<PairedOffice?> = _office.asStateFlow()

    private val _overview = MutableStateFlow<Overview?>(null)
    val overview: StateFlow<Overview?> = _overview.asStateFlow()

    private val _peers = MutableStateFlow<List<Peer>?>(null)
    val peers: StateFlow<List<Peer>?> = _peers.asStateFlow()

    private val _online = MutableStateFlow(true)
    val online: StateFlow<Boolean> = _online.asStateFlow()

    private val _offlineReason = MutableStateFlow<String?>(null)
    val offlineReason: StateFlow<String?> = _offlineReason.asStateFlow()

    private val _pairError = MutableStateFlow<String?>(null)
    val pairError: StateFlow<String?> = _pairError.asStateFlow()

    private val _toast = MutableStateFlow<String?>(null)
    val toast: StateFlow<String?> = _toast.asStateFlow()

    private val _busy = MutableStateFlow(false)
    val busy: StateFlow<Boolean> = _busy.asStateFlow()

    /** nil = la oficina emparejada; si no, el id de otra oficina vista a través de ella. */
    private val _viewing = MutableStateFlow<String?>(null)
    val viewing: StateFlow<String?> = _viewing.asStateFlow()

    private val _panel = MutableStateFlow<PanelState?>(null)
    val panel: StateFlow<PanelState?> = _panel.asStateFlow()

    private val _panelDenied = MutableStateFlow<String?>(null)
    val panelDenied: StateFlow<String?> = _panelDenied.asStateFlow()

    /** A quién le escribe el redactor ("god" = Michael). Solo de esta oficina. */
    private val _recipient = MutableStateFlow("god")
    val recipient: StateFlow<String> = _recipient.asStateFlow()
    fun setRecipient(id: String) { _recipient.value = id }

    var authorize: suspend (SensitiveAction) -> Boolean = { true }

    private var client: RemoteClient? = null
    private var pollJob: Job? = null

    init {
        // Restaura el emparejamiento, igual que init(defaults:) en Swift.
        try {
            val officeJson = secure?.loadOffice()
            val raw = secure?.loadSessionKey()
            if (officeJson != null && raw != null) {
                val office = MunderJson.decodeFromString<PairedOffice>(officeJson)
                adopt(office, raw)
                val pending = secure?.loadPendingCode()
                _phase.value = if (pending != null) Phase.Waiting(pending) else Phase.Paired
            }
        } catch (e: Exception) {
            // Prefs rotas: arranca sin emparejar, como teléfono nuevo.
        }
    }

    private fun adopt(office: PairedOffice, key: ByteArray) {
        val c = RemoteClient(office, key)
        c.onOfficeChanged = { next ->
            try {
                secure?.saveOffice(MunderJson.encodeToString(next))
            } catch (e: Exception) { }
            _office.value = next.copy()
        }
        client = c
        _office.value = office
    }

    private fun persist(office: PairedOffice) {
        _office.value = office
        try {
            secure?.saveOffice(MunderJson.encodeToString(office))
        } catch (e: Exception) { }
    }

    // ── emparejar ──

    fun pair(addressInput: String, deviceName: String) {
        viewModelScope.launch {
            _pairError.value = null
            val address = Address.normalize(addressInput)
            if (address == null) {
                _pairError.value = "Escribe la dirección de tu computadora, por ejemplo 192.168.1.64 o 100.101.4.7."
                return@launch
            }
            _busy.value = true
            try {
                val name = deviceName.trim().ifEmpty { "Android" }
                val r = pairOffice(address, name)
                secure?.saveSessionKey(r.key)
                persist(r.office)
                secure?.savePendingCode(r.code)
                adopt(r.office, r.key)
                _phase.value = Phase.Waiting(r.code)
            } catch (e: RemoteError) {
                _pairError.value = e.message
            } catch (e: Exception) {
                _pairError.value = e.message ?: "No se pudo emparejar"
            } finally {
                _busy.value = false
            }
        }
    }

    /** Pregunta `hello` hasta que un humano acepta el código en la compu. */
    fun waitForAccept() {
        pollJob?.cancel()
        pollJob = viewModelScope.launch {
            while (isActive) {
                val ph = _phase.value
                if (ph !is Phase.Waiting) break
                try {
                    val hello: Hello = client()?.call("hello") ?: break
                    client()?.learn(hello.addresses ?: emptyList())
                    secure?.clearPendingCode()
                    _phase.value = Phase.Paired
                    _toast.value = "¡Listo! Celular emparejado"
                    refresh()
                    break
                } catch (e: RemoteError) {
                    if (e.code == "unknown_device") {
                        _pairError.value = "La computadora ya no tiene esta solicitud (caduca a los 10 minutos, o se rechazó). Empieza de nuevo."
                        break
                    }
                    // waiting / sin alcance todavía: sigue preguntando
                } catch (e: Exception) { }
                delay(2000)
            }
        }
    }

    /** Demo sin red para capturas: misma oficina de ejemplo que iOS. */
    fun loadDemo(overviewJson: String, peersJson: String, panelJson: String? = null) {
        client = null
        _office.value = PairedOffice(
            officeId = "fa801f7ab6693f03",
            name = "michael-victus",
            boxPub = "",
            deviceId = "9ed00cd88840bed1",
            deviceName = "Android de Danny",
            addresses = mutableListOf("192.168.1.64:47831", "100.101.4.7:47831"),
            via = mutableMapOf("192.168.1.64:47831" to "lan", "100.101.4.7:47831" to "tailscale"),
        )
        try {
            _overview.value = MunderJson.decodeFromString<Overview>(overviewJson)
        } catch (e: Exception) { }
        try {
            _peers.value = MunderJson.decodeFromString<PeersReply>(peersJson).peers
        } catch (e: Exception) { }
        if (panelJson != null) {
            try {
                _panel.value = MunderJson.decodeFromString<PanelState>(panelJson)
            } catch (e: Exception) { }
        }
        _online.value = true
        _phase.value = Phase.Paired
    }

    fun forget() {
        _viewing.value = null
        _recipient.value = "god"
        secure?.deleteSessionKey()
        secure?.clearOffice()
        client = null
        _office.value = null
        _overview.value = null
        _peers.value = null
        _pairError.value = null
        _panel.value = null
        _panelDenied.value = null
        _phase.value = Phase.Unpaired
    }

    // ── la oficina ──

    fun view(officeId: String?) {
        if (officeId == _viewing.value) return
        viewModelScope.launch {
            _viewing.value = officeId
            _recipient.value = "god"
            _overview.value = null
            refresh()
        }
    }

    fun refresh() {
        viewModelScope.launch {
            val c = client ?: return@launch
            if (_phase.value !is Phase.Paired) return@launch
            val want = _viewing.value
            try {
                val args = if (want != null) mapOf("office" to want) else emptyMap()
                val next: Overview = c.call("overview", args)
                if (want != _viewing.value) return@launch // cambió mientras volaba
                _overview.value = next
                _online.value = true
                _offlineReason.value = null
            } catch (e: RemoteError) {
                _online.value = false
                _offlineReason.value = when (e.code) {
                    "unknown_device" -> "La oficina ya no reconoce este celular (lo olvidaron en la computadora). Olvídala aquí y vuelve a emparejar."
                    "stale" -> "La hora del celular y la de la computadora no coinciden. Activa la hora automática en los dos."
                    else -> e.message
                }
            } catch (e: Exception) {
                _online.value = false
                _offlineReason.value = e.message
            }
        }
    }

    fun refreshAddresses() {
        viewModelScope.launch {
            val c = client ?: return@launch
            if (_phase.value !is Phase.Paired) return@launch
            try {
                val hello: Hello = c.call("hello")
                c.learn(hello.addresses ?: emptyList())
            } catch (e: Exception) { }
        }
    }

    fun refreshPeers() {
        viewModelScope.launch {
            val c = client ?: return@launch // en demo se quedan los incluidos
            try {
                val r: PeersReply = c.call("peers")
                _peers.value = r.peers
            } catch (e: Exception) {
                if (_peers.value == null) _peers.value = emptyList()
            }
        }
    }

    // ── el estrato Panel ──

    fun refreshPanel() {
        viewModelScope.launch {
            val c = client ?: return@launch
            if (_phase.value !is Phase.Paired) return@launch
            try {
                _panel.value = c.call("panel.state")
                _panelDenied.value = null
            } catch (e: RemoteError) {
                if (e.code == "no_authority") {
                    _panel.value = null
                    _panelDenied.value = e.message
                }
                // La oficina alcanza pero el panel no: se queda el último estado.
            } catch (e: Exception) { }
        }
    }

    fun panelAction(action: String, args: Map<String, String> = emptyMap(), activity: FragmentActivity? = null) {
        if (panelRemoteBlocked.contains(action)) return
        viewModelScope.launch {
            _busy.value = true
            try {
                if (!authorize(SensitiveAction.PANEL)) return@launch
                val c = client ?: return@launch
                val r: PanelActionReply = c.call("panel.action", mapOf("action" to action, "args" to args))
                _toast.value = if (r.text.isEmpty()) (if (r.ok) "Listo" else "No se pudo") else r.text
                refreshPanel()
            } catch (e: Exception) {
                _toast.value = e.message
            } finally {
                _busy.value = false
            }
        }
    }

    /**
     * Dos botones son solo de escritorio por diseño (ver PANEL_OFF en lib-remote.cjs):
     * el móvil no instala un acceso directo del escritorio ni amplía su propia autoridad.
     */
    companion object {
        val panelRemoteBlocked: Set<String> = setOf("shortcut.install", "link.phoneAuthority")
    }

    fun answer(task: TaskItem, text: String) {
        viewModelScope.launch {
            val q = task.question?.q ?: return@launch
            if (!authorize(SensitiveAction.ANSWER)) return@launch
            _busy.value = true
            try {
                val c = client ?: return@launch
                c.callRaw("answer", mapOf("task_id" to task.id, "q" to q, "text" to text))
                _toast.value = "Respuesta enviada a Michael"
                refresh()
            } catch (e: Exception) {
                _toast.value = e.message
                if ((e as? RemoteError)?.code == "question_changed") refresh()
            } finally {
                _busy.value = false
            }
        }
    }

    /** Esta oficina: a Michael o a un agente. Otra oficina: tarea para SU Michael. */
    fun ask(text: String) {
        val peer = _viewing.value
        if (peer != null) {
            delegate(peer, text)
            return
        }
        viewModelScope.launch {
            if (!authorize(SensitiveAction.ASK)) return@launch
            val to = _recipient.value
            _busy.value = true
            try {
                val c = client ?: return@launch
                val args = mutableMapOf<String, Any?>("text" to text)
                if (to != "god") args["agent"] = to
                val r: MessageReply = c.call("ask", args)
                _toast.value = "Enviado a ${r.name ?: "Michael"}"
                _busy.value = false
                refresh()
            } catch (e: Exception) {
                _toast.value = e.message
                _busy.value = false
            }
        }
    }

    fun delegate(peer: String, text: String) {
        viewModelScope.launch {
            if (!authorize(SensitiveAction.DELEGATE)) return@launch
            _busy.value = true
            try {
                val c = client ?: return@launch
                val r: DelegateReply = c.call("delegate", mapOf("office" to peer, "text" to text))
                _toast.value = "Delegada a ${r.office}"
                refresh()
            } catch (e: Exception) {
                _toast.value = e.message
            } finally {
                _busy.value = false
            }
        }
    }

    /** Agrega una dirección a mano (p. ej. la de Tailscale) y prueba que sí llega a ESTA oficina. */
    fun addAddress(input: String) {
        viewModelScope.launch {
            val c = client ?: return@launch
            val office = _office.value ?: return@launch
            val address = Address.normalize(input)
            if (address == null) {
                _toast.value = "Esa dirección no se entiende"
                return@launch
            }
            if (!authorize(SensitiveAction.ADD_ADDRESS)) return@launch
            val key = secure?.loadSessionKey() ?: return@launch
            val probe = RemoteClient(
                PairedOffice(office.officeId, office.name, office.boxPub, office.deviceId, office.deviceName, mutableListOf(address)),
                key,
            )
            try {
                probe.call<Hello>("hello")
                c.add(address, if (Address.looksTailscale(address)) "tailscale" else "manual")
                _toast.value = "Dirección agregada"
            } catch (e: Exception) {
                _toast.value = "Esa dirección no contestó como tu oficina: ${e.message}"
            }
        }
    }

    fun removeAddress(address: String) {
        viewModelScope.launch {
            if ((_office.value?.addresses?.size ?: 0) <= 1) return@launch
            if (!authorize(SensitiveAction.REMOVE_ADDRESS)) return@launch
            client?.remove(address)
        }
    }

    /** Olvida una oficina emparejada, tras comprobar que eres el dueño. */
    fun forgetAuthorized() {
        viewModelScope.launch {
            if (!authorize(SensitiveAction.FORGET)) return@launch
            forget()
        }
    }

    fun toastShown() {
        _toast.value = null
    }

    private fun client(): RemoteClient? = client

    private suspend fun authorize(action: SensitiveAction): Boolean = authorize(action)
}
